# Fake News Classification - EDA

In [1]:
import os
import zipfile
import pandas as pd
import numpy as np

In [2]:
DATA_PATH = '../data/'
# if the path or names of the data hosted on Kaggle changes the following 
# three constants will need updated
DATA_ARCHIVE = 'fake-and-real-news-dataset.zip'
FAKE_DATA_FILE = 'Fake.csv'
TRUE_DATA_FILE = 'True.csv'
KAGGLE_DATA_LOCATION = 'clmentbisaillon/fake-and-real-news-dataset'

# Download the Data

In [3]:
try: 
    os.mkdir(DATA_PATH)
except FileExistsError:
    print('Data directory already exists')
    pass

Data directory already exists


Documentation on the Kaggle API is located at [https://www.kaggle.com/docs/api](https://www.kaggle.com/docs/api)

In [4]:
# assumes that kaggle has been installed and an api key is correctly installed
if not os.path.exists(DATA_PATH + DATA_ARCHIVE):
    !kaggle datasets download -d $KAGGLE_DATA_LOCATION -p $DATA_PATH

In [5]:
if not os.path.exists(DATA_PATH + FAKE_DATA_FILE):
    with zipfile.ZipFile(DATA_PATH + DATA_ARCHIVE, 'r') as zip_ref:
        zip_ref.extractall(path=DATA_PATH)

In [6]:
os.listdir('../data')

['fake-and-real-news-dataset.zip', 'Fake.csv', 'True.csv']

# Load the Data

In [7]:
fake_df = pd.read_csv(DATA_PATH + FAKE_DATA_FILE)

In [8]:
fake_df.head()

title  \
0   Donald Trump Sends Out Embarrassing New Year’...   
1   Drunk Bragging Trump Staffer Started Russian ...   
2   Sheriff David Clarke Becomes An Internet Joke...   
3   Trump Is So Obsessed He Even Has Obama’s Name...   
4   Pope Francis Just Called Out Donald Trump Dur...   

                                                text subject  \
0  Donald Trump just couldn t wish all Americans ...    News   
1  House Intelligence Committee Chairman Devin Nu...    News   
2  On Friday, it was revealed that former Milwauk...    News   
3  On Christmas day, Donald Trump announced that ...    News   
4  Pope Francis used his annual Christmas Day mes...    News   

                date  
0  December 31, 2017  
1  December 31, 2017  
2  December 30, 2017  
3  December 29, 2017  
4  December 25, 2017

In [9]:
true_df = pd.read_csv(DATA_PATH + TRUE_DATA_FILE)

In [10]:
true_df.head()

title  \
0  As U.S. budget fight looms, Republicans flip t...   
1  U.S. military to accept transgender recruits o...   
2  Senior U.S. Republican senator: 'Let Mr. Muell...   
3  FBI Russia probe helped by Australian diplomat...   
4  Trump wants Postal Service to charge 'much mor...   

                                                text       subject  \
0  WASHINGTON (Reuters) - The head of a conservat...  politicsNews   
1  WASHINGTON (Reuters) - Transgender people will...  politicsNews   
2  WASHINGTON (Reuters) - The special counsel inv...  politicsNews   
3  WASHINGTON (Reuters) - Trump campaign adviser ...  politicsNews   
4  SEATTLE/WASHINGTON (Reuters) - President Donal...  politicsNews   

                 date  
0  December 31, 2017   
1  December 29, 2017   
2  December 31, 2017   
3  December 30, 2017   
4  December 29, 2017

In [11]:
true_df['label'] = 'true'

In [12]:
fake_df['label'] = 'fake'

In [13]:
df = pd.concat([fake_df, true_df], ignore_index=True)

In [14]:
df.head()

title  \
0   Donald Trump Sends Out Embarrassing New Year’...   
1   Drunk Bragging Trump Staffer Started Russian ...   
2   Sheriff David Clarke Becomes An Internet Joke...   
3   Trump Is So Obsessed He Even Has Obama’s Name...   
4   Pope Francis Just Called Out Donald Trump Dur...   

                                                text subject  \
0  Donald Trump just couldn t wish all Americans ...    News   
1  House Intelligence Committee Chairman Devin Nu...    News   
2  On Friday, it was revealed that former Milwauk...    News   
3  On Christmas day, Donald Trump announced that ...    News   
4  Pope Francis used his annual Christmas Day mes...    News   

                date label  
0  December 31, 2017  fake  
1  December 31, 2017  fake  
2  December 30, 2017  fake  
3  December 29, 2017  fake  
4  December 25, 2017  fake

# Clean the Data

## Check for missing data

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44898 entries, 0 to 44897
Data columns (total 5 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   title    44898 non-null  object
 1   text     44898 non-null  object
 2   subject  44898 non-null  object
 3   date     44898 non-null  object
 4   label    44898 non-null  object
dtypes: object(5)
memory usage: 1.7+ MB


There are no null values, next I will check for any placeholder values.

## Check for placeholder values and duplicates

### Title column

In [16]:
title_value_counts = df.title.value_counts()

In [17]:
title_value_counts[title_value_counts > 1]

Factbox: Trump fills top jobs for his administration                                                                                         14
Factbox: Contenders for senior jobs in Trump's administration                                                                                 8
Highlights: The Trump presidency on April 13 at 9:30 P.M. EDT/0130 GMT on Friday                                                              8
MEDIA IGNORES Time That Bill Clinton FIRED His FBI Director On Day Before Vince Foster Was Found Dead                                         6
Factbox: International reaction to arrest of Reuters reporters in Myanmar                                                                     6
                                                                                                                                             ..
TRUMP WAS RIGHT About CNN Being “Very Fake News”: Federal Judge Rules Against CNN In “FAKE NEWS” Case…May Have Acted With “Actual Malice

There are a lot of titles that are used more than once.  This will need to be reviewed.
First I will look at the most commonly reused title.

In [18]:
df[df.title == 'Factbox: Trump fills top jobs for his administration']

title  \
29701  Factbox: Trump fills top jobs for his administ...   
29875  Factbox: Trump fills top jobs for his administ...   
30096  Factbox: Trump fills top jobs for his administ...   
30118  Factbox: Trump fills top jobs for his administ...   
30136  Factbox: Trump fills top jobs for his administ...   
30167  Factbox: Trump fills top jobs for his administ...   
30186  Factbox: Trump fills top jobs for his administ...   
30220  Factbox: Trump fills top jobs for his administ...   
30236  Factbox: Trump fills top jobs for his administ...   
30289  Factbox: Trump fills top jobs for his administ...   
30330  Factbox: Trump fills top jobs for his administ...   
30380  Factbox: Trump fills top jobs for his administ...   
30480  Factbox: Trump fills top jobs for his administ...   
30489  Factbox: Trump fills top jobs for his administ...   

                                                    text       subject  \
29701  (Reuters) - U.S. President-elect Donald Trump ...  politicsNews   
29875  (Reuters) - U.S. President-elect Donald Trump ...  politicsNews   
30096  (Reuters) - U.S. President-elect Donald Trump ...  politicsNews   
30118  (Reuters) - President-elect Donald Trump will ...  politicsNews   
30136  (Reuters) - President-elect Donald Trump will ...  politicsNews   
30167  (Reuters) - U.S. President-elect Donald Trump ...  politicsNews   
30186  (Reuters) - U.S. President-elect Donald Trump ...  politicsNews   
30220  (Reuters) - U.S. President-elect Donald Trump ...  politicsNews   
30236  (Reuters) - U.S. President-elect Donald Trump ...  politicsNews   
30289  (Reuters) - U.S. President-elect Donald Trump ...  politicsNews   
30330  (Reuters) - U.S. President-elect Donald Trump ...  politicsNews   
30380  (Reuters) - U.S. President-elect Donald Trump ...  politicsNews   
30480  (Reuters) - President-elect Donald Trump has c...  politicsNews   
30489  (Reuters) - U.S. President-elect Donald Trump ...  politicsNews   

                     date label  
29701   January 19, 2017   true  
29875   January 11, 2017   true  
30096  December 21, 2016   true  
30118  December 19, 2016   true  
30136  December 17, 2016   true  
30167  December 15, 2016   true  
30186  December 13, 2016   true  
30220  December 13, 2016   true  
30236  December 12, 2016   true  
30289   December 8, 2016   true  
30330   December 7, 2016   true  
30380   December 5, 2016   true  
30480  November 30, 2016   true  
30489  November 29, 2016   true

In [19]:
df.iloc[29701].text

'(Reuters) - U.S. President-elect Donald Trump will name former Georgia Governor Sonny Perdue as his nominee for secretary of agriculture on Thursday, a senior transition official said on Wednesday. Here is a list of Republican Trump’s selections for top jobs in his administration.  NOTE: Senate confirmation is required for all the posts except national security adviser and White House posts. Tillerson, 64, has spent his entire career at Exxon Mobil Corp, where he rose to chairman and chief executive officer in 2006. A civil engineer by training, the Texan joined the world’s largest publicly traded energy company in 1975 and led several of its operations in the United States as well as in Yemen, Thailand and Russia. As Exxon’s chief executive, he maintained close ties with Moscow and opposed U.S. sanctions against Russia for its incursion into Crimea. Mnuchin, 54, is a successful private equity investor, hedge fund manager and Hollywood financier who spent 17 years at Goldman Sachs Gro

In [20]:
df.iloc[29875].text

'(Reuters) - U.S. President-elect Donald Trump on Wednesday announced he has chosen David Shulkin, who currently heads the Department of Veterans Affairs healthcare system, to head the agency. Here is a list of Republican Trump’s selections for top jobs in his administration.  NOTE: Senate confirmation is required for all the posts except national security adviser and White House posts. Tillerson, 64, has spent his entire career at Exxon Mobil Corp, where he rose to chairman and chief executive officer in 2006. A civil engineer by training, the Texan joined the world’s largest publicly traded energy company in 1975 and led several of its operations in the United States as well as in Yemen, Thailand and Russia. As Exxon’s chief executive, he maintained close ties with Moscow and opposed U.S. sanctions against Russia for its incursion into Crimea. Mnuchin, 54, is a successful private equity investor, hedge fund manager and Hollywood financier who spent 17 years at Goldman Sachs Group Inc

This title seems to be a story title that Reuters reused, while changing the underlying story.

Next I'll make a dataframe with all the data with duplicate titles

In [21]:
duplicate_title_list = set(title_value_counts[title_value_counts > 1].keys())

In [22]:
df_dup_titles = df[df.title.isin(duplicate_title_list)]

In [23]:
df_dup_titles

title  \
1488    McConnell Says He’ll Obstruct ANY Effort To H...   
1535    McConnell Says He’ll Obstruct ANY Effort To H...   
9087   BARBRA STREISAND Gives Up On Dream Of Impeachi...   
9088   WATCH: SENATOR LINDSEY GRAHAM DROPS BOMBSHELL…...   
9089   “CONSERVATIVE GAY GUY” BLASTS Pence’s Aspen Ne...   
...                                                  ...   
44884  U.S. puts more pressure on Pakistan to help wi...   
44886  Trump talks tough on Pakistan's 'terrorist' ha...   
44887  U.S., North Korea clash at U.N. forum over nuc...   
44889  U.S., North Korea clash at U.N. forum over nuc...   
44893  'Fully committed' NATO backs new U.S. approach...   

                                                    text    subject  \
1488   Siding with Trump s Attorney General Jeff Sess...       News   
1535   Siding with Trump s Attorney General Jeff Sess...       News   
9087   Barbra Streisand was an Obama sycophant and on...   politics   
9088   Everyone suspected the sketchy Steele Dossier ...   politics   
9089   It s been said that good fences make good neig...   politics   
...                                                  ...        ...   
44884  WASHINGTON (Reuters) - The United States sugge...  worldnews   
44886  ISLAMABAD (Reuters) - Outlining a new strategy...  worldnews   
44887  GENEVA (Reuters) - North Korea and the United ...  worldnews   
44889  GENEVA (Reuters) - North Korea and the United ...  worldnews   
44893  BRUSSELS (Reuters) - NATO allies on Tuesday we...  worldnews   

                   date label  
1488       May 12, 2017  fake  
1535       May 10, 2017  fake  
9087       Dec 31, 2017  fake  
9088       Dec 31, 2017  fake  
9089       Dec 30, 2017  fake  
...                 ...   ...  
44884  August 21, 2017   true  
44886  August 22, 2017   true  
44887  August 22, 2017   true  
44889  August 22, 2017   true  
44893  August 22, 2017   true  

[11798 rows x 5 columns]

Looking at the above subset of data, I can see that some rows may be complete duplicates (44887 and 44889) and others may be duplicates except for the date (1488 and 1535).

Are there any duplicate rows?

In [24]:
df_dup_titles[df_dup_titles.duplicated()]

title  \
9942   HILLARY TWEETS MESSAGE In Defense Of DACA…OOPS...   
11446  FORMER DEMOCRAT WARNS Young Americans: “Rioter...   
14925  [VIDEO] #BlackLivesMatter Terrorists Storm Dar...   
23926  Senate tax bill stalls on deficit-focused 'tri...   
24259  Trump warns 'rogue regime' North Korea of grav...   
...                                                  ...   
44709  France unveils labor reforms in first step to ...   
44744  Guatemala top court sides with U.N. graft unit...   
44771  Europeans, Africans agree renewed push to tack...   
44834  Thailand's ousted PM Yingluck has fled abroad:...   
44889  U.S., North Korea clash at U.N. forum over nuc...   

                                                    text       subject  \
9942   No time to waste   we've got to fight with eve...      politics   
11446   Who is silencing political speech, physically...      politics   
14925  They were probably just looking for a  safe sp...      politics   
23926  WASHINGTON (Reuters) - The U.S. Senate on Thur...  politicsNews   
24259  BEIJING (Reuters) - U.S. President Donald Trum...  politicsNews   
...                                                  ...           ...   
44709  PARIS (Reuters) - French President Emmanuel Ma...     worldnews   
44744  GUATEMALA CITY (Reuters) - Guatemala s top cou...     worldnews   
44771  PARIS (Reuters) - Europe s  big four  continen...     worldnews   
44834  BANGKOK (Reuters) - Ousted Thai prime minister...     worldnews   
44889  GENEVA (Reuters) - North Korea and the United ...     worldnews   

                     date label  
9942          Sep 9, 2017  fake  
11446        Mar 10, 2017  fake  
14925        Nov 16, 2015  fake  
23926  November 30, 2017   true  
24259   November 8, 2017   true  
...                   ...   ...  
44709    August 31, 2017   true  
44744    August 29, 2017   true  
44771    August 28, 2017   true  
44834    August 25, 2017   true  
44889    August 22, 2017   true  

[209 rows x 5 columns]

There are 209 duplicate rows that will need to be deleted.

I want to also check for other duplicate situations.

Check for rows with duplicate text

In [25]:
df_dup_titles[df_dup_titles.duplicated(['text'])]

title  \
9114   WHY THIS BLUE-COLLAR DEMOCRAT STRONGHOLD Count...   
9738   LEFTIST STORE OWNER Makes Video Asking Libs To...   
9942   HILLARY TWEETS MESSAGE In Defense Of DACA…OOPS...   
11041  Joe Scarborough BERATES Mika Brzezinski Over “...   
11190  WATCH TUCKER CARLSON Scorch Sanctuary City May...   
...                                                  ...   
44709  France unveils labor reforms in first step to ...   
44744  Guatemala top court sides with U.N. graft unit...   
44771  Europeans, Africans agree renewed push to tack...   
44834  Thailand's ousted PM Yingluck has fled abroad:...   
44889  U.S., North Korea clash at U.N. forum over nuc...   

                                                    text    subject  \
9114   AP News  The regulars amble in before dawn and...   politics   
9738   Things didn t go as well as Nicholle had hoped...   politics   
9942   No time to waste   we've got to fight with eve...   politics   
11041                                                      politics   
11190                                                      politics   
...                                                  ...        ...   
44709  PARIS (Reuters) - French President Emmanuel Ma...  worldnews   
44744  GUATEMALA CITY (Reuters) - Guatemala s top cou...  worldnews   
44771  PARIS (Reuters) - Europe s  big four  continen...  worldnews   
44834  BANGKOK (Reuters) - Ousted Thai prime minister...  worldnews   
44889  GENEVA (Reuters) - North Korea and the United ...  worldnews   

                   date label  
9114       Dec 27, 2017  fake  
9738        Oct 6, 2017  fake  
9942        Sep 9, 2017  fake  
11041      Apr 26, 2017  fake  
11190       Apr 6, 2017  fake  
...                 ...   ...  
44709  August 31, 2017   true  
44744  August 29, 2017   true  
44771  August 28, 2017   true  
44834  August 25, 2017   true  
44889  August 22, 2017   true  

[5964 rows x 5 columns]

There are 5964 rows that have duplicate text.

I see some rows do not have any text, and that will have to be investigated later. 

Check for rows that have duplicate title and text

In [26]:
df_dup_titles[df_dup_titles.duplicated(['title', 'text'])]

title  \
9942   HILLARY TWEETS MESSAGE In Defense Of DACA…OOPS...   
11446  FORMER DEMOCRAT WARNS Young Americans: “Rioter...   
14925  [VIDEO] #BlackLivesMatter Terrorists Storm Dar...   
15892  HOUSE INTEL Slaps Subpoenas on McCain Institut...   
15893  PRICELESS! WATCH MSNBC HOST’S Shocked Response...   
...                                                  ...   
44709  France unveils labor reforms in first step to ...   
44744  Guatemala top court sides with U.N. graft unit...   
44771  Europeans, Africans agree renewed push to tack...   
44834  Thailand's ousted PM Yingluck has fled abroad:...   
44889  U.S., North Korea clash at U.N. forum over nuc...   

                                                    text          subject  \
9942   No time to waste   we've got to fight with eve...         politics   
11446   Who is silencing political speech, physically...         politics   
14925  They were probably just looking for a  safe sp...         politics   
15892  Please see our previous report below on the Mc...  Government News   
15893  THIS IS PRICELESS! The video below shows just ...  Government News   
...                                                  ...              ...   
44709  PARIS (Reuters) - French President Emmanuel Ma...        worldnews   
44744  GUATEMALA CITY (Reuters) - Guatemala s top cou...        worldnews   
44771  PARIS (Reuters) - Europe s  big four  continen...        worldnews   
44834  BANGKOK (Reuters) - Ousted Thai prime minister...        worldnews   
44889  GENEVA (Reuters) - North Korea and the United ...        worldnews   

                   date label  
9942        Sep 9, 2017  fake  
11446      Mar 10, 2017  fake  
14925      Nov 16, 2015  fake  
15892      Dec 27, 2017  fake  
15893      Dec 26, 2017  fake  
...                 ...   ...  
44709  August 31, 2017   true  
44744  August 29, 2017   true  
44771  August 28, 2017   true  
44834  August 25, 2017   true  
44889  August 22, 2017   true  

[5793 rows x 5 columns]

Only 5793 rows with duplicate title and text.

Of the 171 rows that have duplicate text but not title, how many have empty text values

In [27]:
df_dup_titles[df_dup_titles.text == ' ']

title text    subject  \
10923  TAKE OUR POLL: Who Do You Think President Trum...        politics   
11041  Joe Scarborough BERATES Mika Brzezinski Over “...        politics   
11190  WATCH TUCKER CARLSON Scorch Sanctuary City May...        politics   
11236  SHOCKER: Public School Turns Computer Lab Into...        politics   
11247  MICHAEL FLYNN’S LAWYER Releases Statement Scor...        politics   
...                                                  ...  ...        ...   
21816  BALTIMORE BURNS: MARYLAND GOVERNOR BRINGS IN N...       left-news   
21826  FULL VIDEO: THE BLOCKBUSTER INVESTIGATION INTO...       left-news   
21827  (VIDEO) HILLARY CLINTON: RELIGIOUS BELIEFS MUS...       left-news   
21857  (VIDEO)ICE PROTECTING OBAMA: WON’T RELEASE NAM...       left-news   
21873  (VIDEO) HYSTERICAL SNL TAKE ON HILLARY’S ANNOU...       left-news   

               date label  
10923  May 10, 2017  fake  
11041  Apr 26, 2017  fake  
11190   Apr 6, 2017  fake  
11236   Apr 1, 2017  fake  
11247  Mar 30, 2017  fake  
...             ...   ...  
21816  Apr 27, 2015  fake  
21826  Apr 25, 2015  fake  
21827  Apr 25, 2015  fake  
21857  Apr 14, 2015  fake  
21873  Apr 12, 2015  fake  

[349 rows x 5 columns]

There are more rows with a text value of ' ' than 171, so this issue will need to be dealt with separately.

*Summary for title column* There are close to 11,800 rows that have titles used multiple times.  The use of a duplicate title seems to be a common practice and some articles seem to be revisions.  A duplicate title will be considered ok, but the duplicate rows and rows with duplicate text for the story will need removed.  Rows without text will also need to be removed.

### text column

Check for a text value with one space

In [28]:
df[df.text == ' ']

title text       subject  \
10923  TAKE OUR POLL: Who Do You Think President Trum...           politics   
11041  Joe Scarborough BERATES Mika Brzezinski Over “...           politics   
11190  WATCH TUCKER CARLSON Scorch Sanctuary City May...           politics   
11225  MAYOR OF SANCTUARY CITY: Trump Trying To Make ...           politics   
11236  SHOCKER: Public School Turns Computer Lab Into...           politics   
...                                                  ...  ...           ...   
21826  FULL VIDEO: THE BLOCKBUSTER INVESTIGATION INTO...          left-news   
21827  (VIDEO) HILLARY CLINTON: RELIGIOUS BELIEFS MUS...          left-news   
21857  (VIDEO)ICE PROTECTING OBAMA: WON’T RELEASE NAM...          left-news   
21873  (VIDEO) HYSTERICAL SNL TAKE ON HILLARY’S ANNOU...          left-news   
32451                     Graphic: Supreme Court roundup       politicsNews   

                 date label  
10923    May 10, 2017  fake  
11041    Apr 26, 2017  fake  
11190     Apr 6, 2017  fake  
11225     Apr 2, 2017  fake  
11236     Apr 1, 2017  fake  
...               ...   ...  
21826    Apr 25, 2015  fake  
21827    Apr 25, 2015  fake  
21857    Apr 14, 2015  fake  
21873    Apr 12, 2015  fake  
32451  June 16, 2016   true  

[627 rows x 5 columns]

There could be an arbritray amount of whitespace so I will replace all text values that are only whitespace with Nan

In [29]:
df = df.replace(r'^\s*$', np.nan, regex=True)

In [30]:
df.isna().sum()

title        0
text       631
subject      0
date         0
label        0
dtype: int64

*Summary for the text column* 631 rows with no text is small compared to the size of the dataset, so they should be dropped.

### Remove Duplicate Data

In [41]:
len(df)

44898

Drop Duplicate Rows (keeping the first instance)

In [31]:
df_clean = df.drop_duplicates(ignore_index=True)

In [32]:
len(df_clean)

44689

Drop Rows with No Text (just a title)

In [33]:
df_clean = df_clean.dropna()

In [34]:
len(df_clean)

44058

Drop Rows with same title and text (keeping the first instance)

In [35]:
df_clean = df_clean.drop_duplicates(['title', 'text'], ignore_index=True)

In [36]:
len(df_clean)

38658

Drop Rows with the same text (keeping the first instance)

In [37]:
df_clean = df_clean.drop_duplicates(['text'], ignore_index=True)

In [38]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 38644 entries, 0 to 38643
Data columns (total 5 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   title    38644 non-null  object
 1   text     38644 non-null  object
 2   subject  38644 non-null  object
 3   date     38644 non-null  object
 4   label    38644 non-null  object
dtypes: object(5)
memory usage: 1.5+ MB


How many rows were dropped?

In [39]:
len(df) - len(df_clean)

6254

## Check Date Range

### Check date field for a valid format

In [40]:
try:
    df_clean['date'] = pd.to_datetime(df_clean['date'])
except ValueError:
    print('Not all date values are dates.')

Not all date values are dates.


### Filter instances with invalid dates

I will need to filter out the date values that are not dates.

In [43]:
def not_a_date(string):
    """
    Check if the input string can be converted to a date.
    
    Return:
    True if it is not a date
    False if it is a date
    """
    try:
        date = pd.to_datetime(string)
        return False
    except ValueError:
        return True

In [44]:
df_bad_dates = df_clean[df_clean['date'].apply(not_a_date)]

In [45]:
df_bad_dates

title  \
9357   https://100percentfedup.com/served-roy-moore-v...   
15119  https://100percentfedup.com/video-hillary-aske...   
15120  https://100percentfedup.com/12-yr-old-black-co...   
15422  https://fedup.wpengine.com/wp-content/uploads/...   
15423  https://fedup.wpengine.com/wp-content/uploads/...   
16126                                           Homepage   

                                                    text    subject  \
9357   https://100percentfedup.com/served-roy-moore-v...   politics   
15119  https://100percentfedup.com/video-hillary-aske...   politics   
15120  https://100percentfedup.com/12-yr-old-black-co...   politics   
15422  https://fedup.wpengine.com/wp-content/uploads/...   politics   
15423  https://fedup.wpengine.com/wp-content/uploads/...   politics   
16126  [vc_row][vc_column width= 1/1 ][td_block_trend...  left-news   

                                                    date label  
9357   https://100percentfedup.com/served-roy-moore-v...  fake  
15119  https://100percentfedup.com/video-hillary-aske...  fake  
15120  https://100percentfedup.com/12-yr-old-black-co...  fake  
15422  https://fedup.wpengine.com/wp-content/uploads/...  fake  
15423  https://fedup.wpengine.com/wp-content/uploads/...  fake  
16126  MSNBC HOST Rudely Assumes Steel Worker Would N...  fake

The instances with text in the date field are all invalid data and will be dropped

### Drop Instances with invalid dates

In [ ]:
df_clean = df_clean.drop(df_bad_dates.index)

### Convert dates to datetime type

In [48]:
try:
    df_clean['date'] = pd.to_datetime(df_clean['date'])
except ValueError:
    print('Not all date values are dates.')

### Check Date Range

In [49]:
df_clean.date.describe()

count                   38638
unique                   1010
top       2017-12-06 00:00:00
freq                      177
first     2015-03-31 00:00:00
last      2018-02-19 00:00:00
Name: date, dtype: object

The description of the data, indicated that it was between 2016 and 2017.  I will need to check the data that is outside it.

In [52]:
df_early = df_clean[df_clean.date < pd.to_datetime('2016-01-01')]
df_early

title  \
14343  EVERY U.S. CITIZEN TAKEN HOSTAGE IN IRAN To Be...   
14344  WATCH FUNNIEST MAN In American Politics Ridicu...   
14345  FBI POSTS $5,000 REWARD For Person Who Committ...   
14346  SWISS ARMY CHIEF WARNS CITIZENS About Explosiv...   
14347  WOW! Sarah Palin’s Stunning AZ Vacation Home G...   
...                                                  ...   
16665  BREAKING: [Video] COLORADO BAKER WHO REFUSED T...   
16666  (VIDEO) PATRIOTS DEMAND REMOVAL OF COMMUNIST FLAG   
16667  BUSTED: [VIDEO] MAN ATTEMPTS TO TAPE “GOTCHA” ...   
16668  [VIDEO] 16 YR OLD ARRESTED For Violent Gang Be...   
16669  “Non-violence hasn’t worked”…Reverend Sam Most...   

                                                    text    subject  \
14343  Just another slap in the face to US citizens. ...   politics   
14344  You don t want to miss this!Here s a little in...   politics   
14345  Americans were warned by Attorney General, Lor...   politics   
14346  Wouldn t it be great if we had someone in gove...   politics   
14347  It would be great if  her former running mate,...   politics   
...                                                  ...        ...   
16665   About 2,000 of the emails I got were of suppo...  left-news   
16666  Patriots removed the the Chinese flag between ...  left-news   
16667  The Michael Brown and Eric Garner cases have o...  left-news   
16668  This is a sad commentary on a generation who h...  left-news   
16669  Yeah that whole  taking up arms  thing seems t...  left-news   

            date label  
14343 2015-12-31  fake  
14344 2015-12-31  fake  
14345 2015-12-31  fake  
14346 2015-12-31  fake  
14347 2015-12-30  fake  
...          ...   ...  
16665 2015-04-07  fake  
16666 2015-04-06  fake  
16667 2015-04-05  fake  
16668 2015-04-04  fake  
16669 2015-04-01  fake  

[1597 rows x 5 columns]

In [53]:
df_early.label.value_counts()

fake    1597
Name: label, dtype: int64

In [54]:
df_late = df_clean[df_clean.date > pd.to_datetime('2017-12-31')]

In [55]:
df_late

title  \
9050  Democrat Senator Warns Mueller Not To Release ...   
9051  MSNBC ANCHOR Flabbergasted at What Texas Teach...   
9052  WATCH: SNOWFLAKES ASKED Communist Party Platfo...   
9053  JUST IN: BADASS GENERAL JOHN KELLY Shoved Chin...   
9054  DOJ’s JEFF SESSIONS Opens Investigation Into W...   
9055  WATCH: ’’Carpetbagger” Mitt Romney Takes Swipe...   
9056  GOTCHA! A Stammering John Podesta Stunned by a...   
9057  HIGH SCHOOL TEACHER Makes “Shooting At Trump” ...   
9058  IT BEGINS…RINO MEGA-DONOR Threatens Jeb Bush: ...   
9059  BUSTED! Maxine Waters’ Ties to Black Nationali...   
9060  LOCAL REPORTER IN DEEP BLUE STATE Stuns Libera...   
9061  FACEBOOK RELEASES Ads Paid For By Russians…And...   
9062  FACEBOOK’S VP Of Ads BLASTS MEDIA: Majority Of...   
9063  MEGHAN MCCAIN Cut-Off After Saying the Florida...   
9064  BROTHER OF HERO COACH Breaks Down Over Tribute...   
9065  “THE TRUMP CURSE IS REAL”…Social Media Respond...   
9066  How The Left, The Anti Defamation League and M...   
9067  LOOK WHO Approved Visas For Russian Operatives...   
9068  JUST IN: PRESIDENT TRUMP AND FIRST LADY Make S...   
9069  BREAKING: FL GOV RICK SCOTT Calls for FBI Dire...   
9070  HEY, BERNIE SANDERS: 18 SCHOOL SHOOTINGS Have ...   
9071  TUCKER CARLSON Rips The Left On Gun Control Hy...   
9072  WATCH: Mitt Romney Announces Run for Senate…Ba...   
9073  WATCH SHERIFF’S Honest Response After a School...   
9074  Flashback Video: Susan Rice Says Trump’s Claim...   
9075  POLITICS FIRST: LEFTIST PARENTS Of Former Comb...   
9076  TWO HIGH SCHOOL JROTC Members Tell INCREDIBLE ...   
9077  MEDIA PROMOTES Fake News That Shooter Connecte...   
9078  WHAT’S WRONG WITH PELOSI? Nancy Stutters, Stam...   
9079  “PROUD TRANSGENDER” Democrat Candidate For Con...   
9080  JUST IN: Senate Rejects Immigration Bill…Trump...   
9081  SC STATE REP Who is a Former Cop Proposes Smar...   
9082  HUH? RAPPER JAY-Z Compares Trayvon Martin to M...   
9083  The DIRTY TRUTH About DACA Recipients…Where Th...   
9084  JUST IN: CROOKED DOJ OFFICIAL Didn’t Reveal Th...   

                                                   text   subject       date  \
9050  According to The Hill, Democrat Senator Bob Ca...  politics 2018-02-19   
9051  If we protect every other government building ...  politics 2018-02-19   
9052  Ami Horowitz is fantastic! Check out this man ...  politics 2018-02-19   
9053  Just one more reminder of why President Trump ...  politics 2018-02-18   
9054  Thank goodnesss Jeff Sessions is moving on fin...  politics 2018-02-18   
9055  The GOP presidential candidate who couldn t se...  politics 2018-02-18   
9056  John Podesta is Hillary Clinton s former campa...  politics 2018-02-18   
9057  EAG News  A Wyoming high school is under fire ...  politics 2018-02-18   
9058  A longtime Republican Party donor based in Flo...  politics 2018-02-18   
9059  We reported that Maxine Waters and members of ...  politics 2018-02-18   
9060   We took a very hard look at these numbers, an...  politics 2018-02-17   
9061  There s been a lot of talk about the ads Russi...  politics 2018-02-17   
9062  Facebook s VP of Advertising took to Twitter t...  politics 2018-02-17   
9063  It clear that common sense isn t what these he...  politics 2018-02-17   
9064  The hero football coach who threw himself betw...  politics 2018-02-17   
9065  Two months before the Winter Olympics in Seoul...  politics 2018-02-17   
9066  Immediately after the police identified the su...  politics 2018-02-17   
9067  According to Fox News  A federal grand jury on...  politics 2018-02-16   
9068   President Trump visits Florida hospital, prai...  politics 2018-02-16   
9069  While the left points the finger at guns, guns...  politics 2018-02-16   
9070  The left has locked onto a stat that 18 school...  politics 2018-02-16   
9071  Tucker Carlson weighed in on the hysteria over...  politics 2018-02-16   
9072  Just what we need another Trump-bashing RINO. ...  politics 2018-02-16   
9073 

In [56]:
df_clean.label.value_counts()

true    21191
fake    17447
Name: label, dtype: int64

### Summary

All the instances outside the 2016 to 2017 range are labeled Fake.  Given that they look similar to the data in that range, and that the number of instances labeled fake has already been reduced because of duplicate data, I will not remove them.